Fix: Remove deprecated Qt5 Core5Compat dependency

mpconley opened the pull request:
#### Brief overview of PR changes/additions

Removes the deprecated Qt5 Core5Compat module and migrates all text encoding functionality to use Qt6's native APIs. This modernizes Mudlet's codebase to be fully Qt6-compliant.

#### Motivation for adding to Mudlet

The Qt5 Core5Compat module is a compatibility layer that Qt deprecated for removal in future versions. By migrating away from it now, we ensure Mudlet remains buildable and maintainable as Qt continues to evolve. This change has no user-facing impact - all existing functionality including multi-byte character encodings (UTF-8, GBK, BIG5, EUC-KR) and custom codepages (CP437, CP667, CP737, CP869, MEDIEVIA) continues to work exactly as before.

#### Other info (issues closed, discussion etc)

Closes #7386

**Testing completed:**
- ✅ All 14 C++ unit tests passing
- ✅ Application builds successfully on macOS
- ✅ Application launches and runs normally
- ✅ Encoding detection and switching verified (UTF-8, ISO-8859-1)
- ✅ Hunspell spell-checking with custom encodings functional

**Technical details:**
- Replaced QTextCodec with QStringConverter/QStringEncoder/QStringDecoder
- Refactored 5 custom codec classes from inheritance to standalone converters
- Created TEncodingHelper utility class for unified encoding API
- Updated 15+ source files across core networking, display, and Lua subsystems
- Removed Core5Compat from all build configurations (CMake & qmake)

mpconley committed:
initial

mpconley committed:
Merge branch 'development' into remove-qt5-core5compat

SlySven commented on src/TEncodingHelper.cpp:
```diff
@@ -0,0 +1,135 @@
+{
+    if (encoding == "M_CP437" || encoding == "CP437") {
+        return TTextCodec_437::toUnicode(bytes);
+    } else if (encoding == "M_CP667" || encoding == "CP667") {
```
Remember the `else-after-return` anti-pattern? Those `else`s are redundant! :grinning:

mpconley committed:
Fix dangling pointer in ctelnet::decodeBytes()

Remove .constData() call that was taking a pointer into a temporary
QByteArray returned by toUtf8(). The temporary was immediately
destroyed, leaving a dangling pointer. Now returns the QByteArray
by value, allowing proper copy/move semantics.

Addresses @coderabbitai review comment:
https://github.com/Mudlet/Mudlet/pull/8550#discussion_r2540219601

SlySven commented on src/TTextCodec.cpp:
```diff
@@ -113,847 +108,344 @@ const QVector<QChar> TTextCodec_869::CptoUnicode{
     QChar(0x00AD), QChar(0x00B1), QChar(0x03C5), QChar(0x03C6), QChar(0x03C7), QChar(0x00A7), QChar(0x03C8), QChar(0x0385),  // F0=F7
     QChar(0x00B0), QChar(0x00A8), QChar(0x03C9), QChar(0x03CB), QChar(0x03B0), QChar(0x03CE), QChar(0x25A0), QChar(0x00A0)}; // F8-FF
 
-// This was based on WINDOWS-1252 but has some additonal codepoints that are
```
Please restore/leave these comments (both the header and the per-codepoint ones) in place - they document a very specific situation that may not be documented anywhere else.

mpconley committed:
Address CodeRabbit review feedback

- Fix TLuaInterpreter Hunspell bugs: spellCheckWord now uses encodedText, spellSuggestWord decodes suggestions
- Fix ctelnet.cpp: use toStdString() instead of constData() to avoid null-byte truncation
- Add explicit QList includes to TTextCodec.h and TEncodingHelper.h for header self-containment
- Remove unused QTextCodec forward declaration from TBuffer.h
- Add codec name validation in TCommandLine spell-checking
- Fix TEncodingHelper::canEncode to return false for unknown encodings
- Remove obsolete mpHunspellCodec_system member from TMainConsole.h

Addresses nitpick and critical comments from CodeRabbit review:
https://github.com/Mudlet/Mudlet/pull/8550#pullrequestreview-3480438447

mpconley committed:
Fix user dictionary suggestion decoding to use UTF-8

User dictionary is always UTF-8 encoded, but suggestions were being
decoded using the system dictionary's encoding (codecName). This caused
mojibake when system dictionary uses non-UTF-8 encoding like ISO-8859-1.

Changed lines 857 and 864 to use QString::fromUtf8() instead of
TEncodingHelper::decode() with system codecName.

Addresses critical issue from CodeRabbit review:
https://github.com/Mudlet/Mudlet/pull/8550#pullrequestreview-3480568821

SlySven reviewed (commented):
:yawning_face: too late for me to finish looking at this - but I am a little worried - I'm not sure that the "state" between processing chunks of data (used to handle gaps in a network text stream which break up a multi-byte codepoint - not an issue for the extended ASCII encodings though, and the handling of adding/skipping a BOM if needed - probably not in our usages) - when I created the `TTextCode_XXX` classes I had to include code for these things when implemented the required methods for them...

mpconley committed:
Address SlySven's review feedback

1. Remove else-after-return anti-pattern in TEncodingHelper.cpp
   - Changed if-else-if chains to sequential if statements
   - Each if returns early, making else clauses redundant

2. Restore Medievia codec documentation in TTextCodec.cpp
   - Header comment explaining WINDOWS-1252 base and Medievia font
   - Per-codepoint comments for all 128 extended ASCII mappings
   - Documents mapping to Private Use Area (U+E100+) for custom glyphs
   - Comments preserved for historical/maintenance purposes

Addresses feedback from:
https://github.com/Mudlet/Mudlet/pull/8550#pullrequestreview-3480491054

mpconley commented:
> 🥱 too late for me to finish looking at this - but I am a little worried - I'm not sure that the "state" between processing chunks of data (used to handle gaps in a network text stream which break up a multi-byte codepoint - not an issue for the extended ASCII encodings though, and the handling of adding/skipping a BOM if needed - probably not in our usages) - when I created the `TTextCode_XXX` classes I had to include code for these things when implemented the required methods for them...

Per my spirit animal, Claude 4.5:

> Re: State management concern - Good catch on thinking about this! The state handling is actually already in place, just at a different layer than the old QTextCodec approach:
>
> Custom codecs (CP437/667/737/869/MEDIEVIA): As you noted, these single-byte encodings don't need state management
>
> Multi-byte encodings (UTF-8/GBK/BIG5/EUC-KR): State IS preserved via TBuffer::mIncompleteSequenceBytes which stores partial multi-byte sequences between packets and prepends them to the next chunk (see TBuffer.cpp:403-408, 4314-4317)
>
> BOM handling: Qt6's QStringDecoder in TEncodingHelper::decode() handles BOM automatically
> 
> The key difference: Instead of each codec maintaining state via ConverterState, the TBuffer layer handles incomplete sequences for ALL encodings uniformly. This actually simplifies the architecture since state management is centralized.
>
> Does this address your concern, or should we add explicit tests for multi-byte sequences split across packets?
>
> No code changes needed - the implementation is already correct! 🎉

vadi2 commented on src/ctelnet.h:
```diff
@@ -367,11 +364,7 @@ public slots:
-    QTextCodec* outgoingDataCodec = nullptr;
-//    QTextDecoder* incomingDataDecoder;
-    QTextEncoder* outgoingDataEncoder = nullptr;
+    // Encoding names (no longer using QTextCodec/QTextEncoder/QTextDecoder)
```
Is this comment right?

vadi2 commented on src/TBuffer.h:
```diff
@@ -568,7 +567,7 @@ class TBuffer
 
     QByteArray mEncoding;
-    QTextCodec* mMainIncomingCodec = nullptr;
+    // No longer storing QTextCodec pointer - using TEncodingHelper instead
```
Think we can remove this

vadi2 reviewed (approved):
Big lift, nice work!

mpconley committed:
Merge branch 'development' into remove-qt5-core5compat

mpconley committed:
Remove unnecessary comments per vadi2's review

Removed outdated comments about QTextCodec removal:
- src/ctelnet.h: Removed 'Encoding names (no longer using...)' comment
- src/TBuffer.h: Removed 'No longer storing QTextCodec pointer...' comment

These comments were temporary explanations during the migration and are
no longer needed now that the code is clear and the PR is well-documented.

Addresses feedback from:
https://github.com/Mudlet/Mudlet/pull/8550#pullrequestreview-3481187833
